### PR TITLE
Simplify GlobalCheckpointService and properly hook it for cluster state updates

### PR DESCRIPTION
**Pull request by bleskes**
During a recent merge from master, we lost the bridge from IndicesClusterStateService to the GlobalCheckpointService of primary shards, notifying them of changes to the current set of active/initilizing shards. This commits add the bridge back (with unit tests). It also simplifies the GlobalCheckpoint tracking to use a simpler model (which makes use the fact that the global check point sync is done periodically).

The old integration CheckpointIT test is moved to IndexLevelReplicationTests. I also added simliar assertions to RelocationsIT, which surfaced a bug in the primary relocation logic and how it plays with global checkpoint updates. The test is currently await-fixed and will be fixed in a follow up issue.

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/index/seqno/GlobalCheckpointTests.java [hunk 1]
```diff
@@ -41,89 +49,77 @@
...
+        for (int i = randomIntBetween(min, max); i > 0; i--) {
+            allocations.put("id_" + aIdGenerator.incrementAndGet(), (long)randomInt(1000));
         }
+        return allocations;
+    }
 
-        if (maxLocalCheckpoint == Long.MAX_VALUE) {
+    public void testGlobalCheckpointUpdate() {
+        Map<String, Long> allocations = new HashMap<>();
+        Map<String, Long> tmpAllocations = randomAllocations(0, 5);
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/index/seqno/GlobalCheckpointTests.java]: Can we call this `tmpActive` and not reuse it for the initializing as well `tmpInitializing`?

File: core/src/test/java/org/elasticsearch/index/seqno/GlobalCheckpointTests.java [hunk 2]
```diff
@@ -140,4 +136,112 @@ public void testGlobalCheckpointUpdate() {
...
     }
+
+    public void testMissingActiveIdsPreventAdvance() {
+        final Map<String, Long> active = randomAllocations(1, 5);
+        final Map<String, Long> initializing = randomAllocations(0, 5);
+        final Map<String, Long> assigned = new HashMap<>();
+        assigned.putAll(active);
+        assigned.putAll(initializing);
+        checkpointService.updateAllocationIdsFromMaster(
+            new HashSet<>(randomSubsetOf(randomInt(active.size()-1), active.keySet())),
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/index/seqno/GlobalCheckpointTests.java]: Nit: `active.size()-1` -> `active.size() - 1`

File: core/src/test/java/org/elasticsearch/index/seqno/GlobalCheckpointTests.java [hunk 3]
```diff
@@ -140,4 +136,112 @@ public void testGlobalCheckpointUpdate() {
...
+        assertTrue(checkpointService.updateCheckpointOnPrimary());
+        assertThat(checkpointService.getCheckpoint(), not(equalTo(UNASSIGNED_SEQ_NO)));
+    }
+
+    public void testMissingInSyncIdsPreventAdvance() {
+        final Map<String, Long> active = randomAllocations(0, 5);
+        final Map<String, Long> initializing = randomAllocations(1, 5);
+        checkpointService.updateAllocationIdsFromMaster(active.keySet(), initializing.keySet());
+        initializing.keySet().forEach(checkpointService::markAllocationIdAsInSync);
+        randomSubsetOf(randomInt(initializing.size()-1),
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/index/seqno/GlobalCheckpointTests.java]: Nit: `initializing.size()-1` -> `initializing.size() - 1`

**Review by jasontedor — COMMENTED** [review]
This is a nice simplification @bleskes. I left some nits.

File: core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java [hunk 4]
```diff
@@ -25,47 +25,41 @@
...
-
     /**
-     * This map holds the last known local checkpoint for every initializing shard copy that is still undergoing recovery.
-     * These shards <strong>do not</strong> participate in determining the global checkpoint. This map is needed to make sure that when
-     * shards are promoted to {@link #inSyncLocalCheckpoints} we use the highest known checkpoint, even if we index concurrently
-     * while recovering the shard.
-     * Keyed by allocation ids
+     * This set hold the last set of known valid allocation ids as received by the master. This is important to make sure
+     * shard that are failed or relocated are cleaned up from {@link #inSyncLocalCheckpoints} and do not hold the global
+     * checkpoint back;
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java]: not semicolon , no enter. :)

File: core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java [hunk 5]
```diff
@@ -83,51 +77,34 @@
...
-        activeLocalCheckpoints = new ObjectLongHashMap<>(1 + indexSettings.getNumberOfReplicas());
-        inSyncLocalCheckpoints = new ObjectLongHashMap<>(indexSettings.getNumberOfReplicas());
-        trackingLocalCheckpoint = new ObjectLongHashMap<>(indexSettings.getNumberOfReplicas());
+        inSyncLocalCheckpoints = new ObjectLongHashMap<>(1 + indexSettings.getNumberOfReplicas());
+        assignedAllocationIds = new HashSet<>(1 + indexSettings.getNumberOfReplicas());
         this.globalCheckpoint = globalCheckpoint;
     }
 
     /**
      * notifies the service of a local checkpoint. if the checkpoint is lower than the currently known one,
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java]: Nit correction badge unlocked.

File: core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java [hunk 6]
```diff
@@ -83,51 +77,34 @@
...
+        if (indexOfKey >= 0) {
+            final long current = inSyncLocalCheckpoints.indexGet(indexOfKey);
+
+            if (current < localCheckpoint) {
+                inSyncLocalCheckpoints.indexReplace(indexOfKey, localCheckpoint);
+                if (logger.isTraceEnabled()) {
+                    logger.trace("updated local checkpoint of [{}] to [{}] (was [{}])", allocationId, localCheckpoint, current);
+                }
+            } else {
+                logger.trace("skipping update local checkpoint [{}], current check point is higher " +
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java]: fixed

File: core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java [hunk 7]
```diff
@@ -181,77 +153,53 @@ synchronized void updateCheckpointOnReplica(long globalCheckpoint) {
...
             throw new IllegalArgumentException("global checkpoint from primary should never decrease. current [" +
                 this.globalCheckpoint + "], got [" + globalCheckpoint + "]");
-
         }
     }
 
     /**
      * Notifies the service of the current allocation ids in the cluster state. This method trims any shards that
-     * have been removed and adds/promotes any active allocations to the {@link #activeLocalCheckpoints}.
+     * have been removed
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java]: fixed

File: core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java [hunk 8]
```diff
@@ -181,77 +153,53 @@ synchronized void updateCheckpointOnReplica(long globalCheckpoint) {
...
-            trackingLocalCheckpoint.put(initID, SequenceNumbersService.UNASSIGNED_SEQ_NO);
-            logger.trace("added [{}] to the tracking map due to a CS update", initID);
-
-        }
+        inSyncLocalCheckpoints.removeAll(key -> assignedAllocationIds.contains(key) == false);
     }
 
     /**
      * marks the allocationId as "in sync" with the primary shard. This should be called at the end of recovery
      * where the primary knows all operation bellow the global checkpoint have been completed on this shard.
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java]: this one I will never learn.

File: core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java [hunk 9]
```diff
@@ -181,77 +153,53 @@ synchronized void updateCheckpointOnReplica(long globalCheckpoint) {
...
      *
      * @param allocationId    allocationId of the recovering shard
-     * @param localCheckpoint the local checkpoint of the shard in question
      */
-    public synchronized void markAllocationIdAsInSync(String allocationId, long localCheckpoint) {
-        if (trackingLocalCheckpoint.containsKey(allocationId) == false) {
+    public synchronized void markAllocationIdAsInSync(String allocationId) {
+        if (assignedAllocationIds.contains(allocationId) == false) {
             // master have change its mind and removed this allocation, ignore.
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java]: fixed.

File: core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java [hunk 10]
```diff
@@ -720,6 +731,15 @@ public void handle(final IndexShard.ShardFailure shardFailure) {
          * @throws IOException                  if shard state could not be persisted
          */
         void updateRoutingEntry(ShardRouting shardRouting) throws IOException;
+
+        /**
+         * Notifies the service of the current allocation ids in the cluster state.
+         * see {@link GlobalCheckpointService#updateAllocationIdsFromMaster(Set, Set)} for details.
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java]: grammar is hard.

File: core/src/test/java/org/elasticsearch/index/replication/ESIndexLevelReplicationTestCase.java [hunk 11]
```diff
@@ -230,9 +256,33 @@ public synchronized void close() throws Exception {
         public IndexShard getPrimary() {
             return primary;
         }
+
+        private void syncGlobalCheckpoint() {
+            PlainActionFuture<ReplicationResponse> listner = new PlainActionFuture<>();
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/replication/ESIndexLevelReplicationTestCase.java]: good catch

File: core/src/test/java/org/elasticsearch/index/seqno/GlobalCheckpointTests.java [hunk 12]
```diff
@@ -41,89 +49,77 @@
...
+        Map<String, Long> allocations = new HashMap<>();
+        Map<String, Long> tmpAllocations = randomAllocations(0, 5);
+        Set<String> active = new HashSet<>(tmpAllocations.keySet());
+        allocations.putAll(tmpAllocations);
+        tmpAllocations = randomAllocations(0, 5);
+        Set<String> initializing = new HashSet<>(tmpAllocations.keySet());
+        allocations.putAll(tmpAllocations);
+        assertThat(allocations.size(), equalTo(active.size() + initializing.size()));
+        final long maxLocalCheckpoint;
+        if (allocations.isEmpty()) {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/seqno/GlobalCheckpointTests.java]: ++

File: core/src/test/java/org/elasticsearch/index/seqno/GlobalCheckpointTests.java (hunk 1 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/seqno/GlobalCheckpointTests.java]: agreed this is confusing. Fixed.

**Comment by bleskes** [comment]
Thx @jasontedor . I address all your comments. can you take a second look?

File: core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java (hunk 10 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java]: Computers are harder.

File: core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java [hunk 13]
```diff
@@ -25,47 +25,41 @@
...
+import static org.elasticsearch.index.seqno.SequenceNumbersService.UNASSIGNED_SEQ_NO;
+
 /**
  * A shard component that is responsible of tracking the global checkpoint. The global checkpoint
  * is the highest seq_no for which all lower (or equal) seq_no have been processed on all shards that
  * are currently active. Since shards count as "active" when the master starts them, and before this primary shard
- * has been notified of this fact, we also include shards in that are in the
- * {@link org.elasticsearch.index.shard.IndexShardState#POST_RECOVERY} state when checking for global checkpoint advancement.
- * We call these shards "in sync" with all operations on the primary (see {@link #inSyncLocalCheckpoints}.
+ * has been notified of this fact, we also include shards in that have completed recovery. These shards have received
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java]: Can you fix this one?

**Review by jasontedor — CHANGES_REQUESTED** [review]
Back to you. :smile:

File: core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java (hunk 13 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java]: yep. missed it. it is fixed now.

File: core/src/test/java/org/elasticsearch/index/seqno/GlobalCheckpointTests.java (hunk 2 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/seqno/GlobalCheckpointTests.java]: added

**Comment by bleskes** [comment]
@jasontedor thx. I pushed two more commits.

**Review by jasontedor — APPROVED** [review]
LGTM.

**Comment by bleskes** [comment]
thx @jasontedor

**Pull request merged**